subprojects/docs/src/docs/userguide/tasks.xml

```diff
@@ -290,52 +290,673 @@
+                            <td colspan="3"><emphasis role="bold">*</emphasis> In fact, <literal>File</literal>
+                                can be any type accepted by <literal>Project.file()</literal> and
+                                <literal>FileCollection</literal> can be any type accepted by
+                                <literal>Project.files()</literal>. This includes instances of
+                                <literal>Callable</literal>, such as closures, allowing for lazy
+                                evaluation of the property values.
+                            </td>
+                        </tr>
+                    </tfoot>
+                </table>
+
+                <para>The last two annotations in the table are a special case as they don’t declare
+                    either task inputs or task outputs. So why use them? Because you can make use of
+                    the Java Gradle Plugin Development plugin to help you develop and publish your own
+                    plugins. And this plugin will check whether any properties of your custom task
+                    classes lack an annotation. This protects you against forgetting to add an appropriate
+                    annotation during development.
+                </para>
+
+                <para>Custom task classes are an easy way to bring your own build logic into the arena
+                    of incremental build, but you don’t always have that option. Perhaps you are working
+                    with a task whose source you can’t modify or with an ad-hoc task. In either case,
+                    you can still take advantage of incremental build by using the <literal>TaskInputs</literal>
+                    and <literal>TaskOutputs</literal> runtime APIs.
+                </para>
+            </section>
+
+            <section id="sec:task_input_output_runtime_api">
+                <title>Runtime API</title>
+                <para>When you don’t have access to the source for a custom task class, there is no way
+                    to add any of the annotations we covered in the previous section. Fortunately, Gradle
+                    provides a runtime API that provides access to the same functionality.
+                </para>
+
+                <section id="sec:runtime_api_for_adhoc">
+                    <title>Using it for ad-hoc tasks</title>
+                    <para>This runtime API is provided through a couple of aptly named properties that
+                        are available on every Gradle task:</para>
+                    <itemizedlist>
+                        <listitem>
+                            <literal>inputs</literal> of type <apilink class="org.gradle.api.tasks.TaskInputs"/>
+                        </listitem>
+                        <listitem>
+                            <literal>outputs</literal> of type <apilink class="org.gradle.api.tasks.TaskOutputs"/>
+                        </listitem>
```

Comment:
The downside of this is that the text will say `Task.getInputs()` rather than `inputs`. Just a forewarning.